### ISSUE TITLE
Release 2.1.0 final

### ISSUE THREAD

**Issue by szczepiq** [mockito/mockito#642]
The purpose of this ticket is to collect all actions and requirements for getting 2.1.0 final. Let's not wait too long :)

- [x] investigate #638 
- [x] identify opportunities to promo the RC and get more feedback
- [x] decide how long do we wait for feedback before publishing final version
- [x] @TimvdLippe review the #648
- [x] #648 - include more info the javadoc, e.g. direction of the feature, what we expect from the user, inform it is incubating, etc
- [x] @szczepiq review the #648
- update the release notes re #648
 - [x] inform about the new feature (incubating, motivating, future direction)
- [x] publish another RC and spread the word

**Comment by TimvdLippe** [comment]
I think we can push `2.1.0` next week since RC has been available for a little while now. Also I submitted a (random) PR to Google Guava and 850k+ tests did not fail :tada: https://github.com/google/guava/pull/2567

**Comment by TimvdLippe** [comment]
Shall I publish a RC tomorrow after merging #646 ?

**Comment by szczepiq** [comment]
-1 for merging #646 unless it's safe with unexpected system properties
-1 for RC for #646. If this is the only bug / improvement, I suggest we push out 2.1.0 final, followed by 2.1.1 that contains the improvement. While I like the improvement, it does not strike me as something that needs another release candidate.

Given that we don't have more feedback to investigate, I suggest to set the final release to Monday. Good day to release :)

**Comment by raphw** [comment]
I think we should merge #646 before releasing a final version. We already got several reported issues that were caused by old Java versions and I expect that quite a lot of people might not have updated their JDK in years.

Otherthan that, I agree that we do not need a release candidate for this.

**Comment by TimvdLippe** [comment]
Oh oops I indeed did not mean RC, but the final release :p

So lets merge #646 when the issue is fixed and then publish `2.1.0` on monday :tada:

**Comment by bric3** [comment]
I will tweak the code before the week-end, I was short on time these past two week.

**Comment by szczepiq** [comment]
The final version needs to be exactly the same as the last RC. That's the idea of "release candidates" :)

Essentially, the options are:

a) publish 2.1.0 final now, include the improvement in 2.1.1 (no rc needed, after 2.1.0 is out we won't be releasing candidates for 2.x versions)
b) wait for fixes in the improvement, merge it, publish 2.1.0-rc-2, wait some time (for example, a day), publish 2.1.0 final

@raphw says it's important to include the fix, so let's do option b)

**Comment by TimvdLippe** [comment]
It is not strictly required for 1 very small bug fix to create a whole new RC with the accompanying waiting time. Other teams have also published their latest RC with just a couple of bug fixes immediately as final release, so I think we can safely release 2.1.0 with #646 on monday.

**Comment by szczepiq** [comment]
The fact that other teams do it does not mean that it's right :)

We want to be in the position that we can always say that: "the RC is exactly the final release". This helps convincing teams to use RCs.

Of course nothing can stop you from publishing final with any number of fixes you want. However, it is simply the matter of good practice to follow version conventions and avoid making exceptions for no good reason.

Publishing final version with patches on top of last RC is sloppy. The reason one might want to do it is when the release process is manual / hard. It's not the case for Mockito, where entire pipeline is automated!!! :D

**Comment by raphw** [comment]
I agree with you that we should **not** push this to the final release but I really think that we should merge #646 before going final. Therefore, I suggest a second release candidate instead of a release on Monday. Let's please do this right. Isn't this what RCs are for? Doing so, we can also merge #648.

The problem addressed by #646 was reported at least 6 times. Many developers do not even mean to run old Java versions but have some outdated JDK bundled with their IDE. This means that tests run using Maven but not in, for example, Eclipse, what is difficult to debug and will most likely cause frustration among some early adopters of Mockito 2. In the end, I have spent some hours myself to investigate reported issues of this origin. If we had thought of something like @bric3's solution earlier, I am sure we would have done this right away and I was very happy to see this PR. I am also sure this automatic debugging help will free us some time once Mockito 2 hits the mainstream and more people run into this problem.

As for #648, I think this can make quite a difference in how the Mockito 2 release is perceived in the public. A lot of people have already implemented their own stream-aware "smart null" answers or Java 8-adapters - thanks to the great Mockito API making this so easy - and most of our other changes are internal clean-ups and and do not really change user experience. With adding support for mocking final types and methods, we are delivering something new and exciting, something that people will speak about. Without such a feature, we are risking to suffer the Java 7-phenomenon where very few people bothered to update despite the significant changes made in the OpenJDK. In the end, to a user updating is a risk and additional work; so why bother if there is nothing in for you? In this perspective, Java 8 was an entirely different story. And *true* final mocks do not really exist in the Java-space yet, in this respect we have built something that was never done before! I am sure people will want to try this feature out what hopefully motivates more folks to an upgrade.

Given that we have 2 PRs that only need to be merged with features that are both helping migration and adding some fuzz about the upcoming release, I am strictly against going for a final version on Monday. Let's pick this low hanging fruit. Also, from a marketing perspective, releasing Mockito 2.1 on Monday and Mockito 2.2-RC on Tuesday, I do not think this would sell well.

I wish I had time to do all this before the RC was published but I was too busy during JavaZone and I had to catch up some sleep before getting everything I wanted ready. But now we have waited so long, let's delay the final version for just a little bit more. Isn't anticipation the greatest joy anyways?

**Comment by szczepiq** [comment]
>The problem addressed by #646 was reported at least 6 times. 

I totally agree. Let's make java8 warning a part of 2.1.0. All I want is to follow good practices of versioning and RC.

>releasing Mockito 2.1 on Monday and Mockito 2.2-RC on Tuesday, I do not think this would sell well.

We don't publish RCs for _compatible_ versions. We don't plan RCs for 2.2, 2.3, 2.12.32, they are compatible versions and don't need RCs.

>And true final mocks do not really exist in the Java-space yet, in this respect we have built something that was never done before!

LOVE IT!!!!

>As for #648, I think this can make quite a difference in how the Mockito 2 release is perceived in the public.

That's really cool :) Caveats:

- it's late
- the feature is not ready (assuming the code is ok, documentation? release notes?)
- the feature is a breaking change due to extra dependencies

I'm not sure why you push so hard to include it in 2.1.0. You can spend more time on this feature, make it great, compatible, well documented, push it with 2.2.0 few weeks from now, include buzz, promo, etc. You will achieve __exactly__ the same results community-wise. Actually, I think it is even better to spread good features over time to demonstrate that Mockito continually improves. Community will be adopting Mockito 2 slowly, it will take years. Some projects will stay on 1.10 forever :)

I'm curious:

- how much time is needed to make the feature is ready?
- how much time is needed to make the feature backwards compatible?

@raphw, I'm very happy you work on this and I totally admire your passion to improve Mockito. Let's build a plan and make it work. This feature is too late for 2.1.0. Last minute, unannounced big features, in the RC phase - this is just not a good idea. We've been planning the release for months, inline mocks never appeared on the checklists.

Long story short, we don't need to block 2.1.0 to make this available in 2.x and make it an awesome, neat feature!!! :)

**Comment by raphw** [comment]
As I said before, the feature is ready: it passes all of Mockito's unit tests when it is activated and we even added this to the Travis build to be constantly verified. At the same time, the feature guarantees to not break anything.

I want to merge this now as there is nothing left to do, this is why the feature is a PR ready to be merged and not a private branch. If it is not included in 2.1, I can live with this but I find this a missed opportunity. From my perspective, this feature is good to go and I do not plan to spend more time on it, even if it is not included in 2.1.

**Comment by bric3** [comment]
Usually following process is is good, but some time following guts is nice too. I second @raphw to include the mockmaker, it's opt-in. 

I agree final mock never appeared on the radar here, but the topic was discussed and investigated more than a year ago, by me and especially Rafael who, in the end, achieved this alone. That's a fantastic work @raphw !!!

Including this is an opportunity. We already included incubating API before, I don't see why we shouldn't include it in 2.1.0 (and as such an 2.1.0-RC.x ) as well.

**Comment by szczepiq** [comment]
Thank you guys challenging me and the good practices. We should be challenging one another and keep growing as technical leaders. Discussions like those are the core of collaborative software engineering. Let's enjoy them and never feel discouraged! I find you guys the best crowd I worked with in a while, and @raphw, you did pull off a feature nobody yet did in the history of mocking :)

About the feature: it's too late to include it in the 2.1.0 final. It's great work and it must be included in Mockito. This feature is incompatible in the current form because of new dependencies. 

I will continue pushing Mockito team to follow decent engineering practices:
 - when RC is published, subsequent RCs are triggered for bugfixes required to get the release out
 - subsequent RCs are not used to ship new features, especially when they include breaking changes
 - no sudden big changes, midnight before major version release ;)
 - favoring incremental improvements, for better stability

Mockito is the product that has 2M users, it deserves thoughtful and rigorous following of decent engineering practices. Of course exceptions are necessary, but for good reasons!

>Effectively, what you suggests means therefore that we release 2.1.Final now and 2.2.Final in a few weeks (?). 2.2.Final will be the exact same jar as 2.1.Final would be with this feature merged. I simply do not see how this would improve anything.

We can merge and publish version 2.2.0 even next day after the release, so long the feature is ready & compatible. I haven't reviewed the PR thoroughly yet, anyone else did? Does it have user-facing documentation (e.g. how can I opt-in for the feature, what can I expect?). Is it utterly and completely safe for default behavior?

**Comment by bric3** [comment]
> Is it utterly and completely safe for default behavior?

**YES**, it uses the plugin loader for that, people that want to try final class mocks will have to explicitly make a file in the folder `mockito-extensions`.

> Does it have user-facing documentation (e.g. how can I opt-in for the feature, what can I expect?)

See [`Mockito` class](https://github.com/mockito/mockito/pull/648/files#diff-ddbc0d47e06a589778a6efe24c5d71d3R1220)

> I haven't reviewed the PR thoroughly yet, anyone else did?

**YES**, I did

> * subsequent RCs are not used to ship new features, especially when they include breaking changes

Adding a dependency is not a breaking change. However for other reasons I think mockito should avoid relying that dependency (I'll look at shadowing the jar or needed classes).

> * no sudden big changes, midnight before major version release ;)

For such a change, actually an opt-in feature, I think we should release another release candidate. (We will have another RC anyway due to the java version warning.)

**Comment by szczepiq** [comment]
@bric3, shouting not needed :) Thank you for the answers! It's really helpful.

The documentation looks nice, I'll make some edits in the branch.

Adding new dependency is backwards incompatible. I will write later why :)

I'm +1 to including it in RC because:
- I'm confident about the quality and documentation
- It's an opt-in feature
- New dependency is shadowed
- The other new dependency (agent) is a part of ByteBuddy, so I'm willing to compromise the backwards compatibility of the RCs this time.

I will write a comment to PR with few things I want to get done before it's merged.

Great work, thank you!

**Comment by szczepiq** [comment]
Adding new dependency is a breaking change: https://www.linkedin.com/pulse/sofware-craftsmanship-adding-new-dependency-breaking-change-faber :)

**Comment by TimvdLippe** [comment]
I don't think I fully agree with your article. In very technical terms, yes it is a breaking change. However, claiming that it breaks builds is imo a stretch in this particular case. The dependency is very small, has almost no new versions and looking at the release schedule, only 3 minor versions were available longer than a week. Therefore the likelihood of having a dependency conflict (even in minor range) is close to zero (if not zero). Therefore I think it is an exaggeration to claim it is something to worry about.

I would also not like to point to your thoughtpiece in the documentation. I do acknowledge your point of view, but I simply do not share it. Therefore I would like to refrain from putting it in the official documentation, since that would indicate it is shared by the whole Mockito dev team.

**Comment by szczepiq** [comment]
The article does not discuss the probability of failure. It simply describes facts. Change that can break clients is by definition "incompatible". The team can still decide to ship such change, as we are doing :)

I respect your POV. Write your own article!

I intend to link to the article. However, I would love to link to your article, too. Mockito needs more traction and publicity, strong opinions and public discussions is good for us all.

I'm serious, write the article!

**Comment by ChristianSchwarz** [comment]
@szczepiq 
I don't see a problem of adding a new dependency since the major version of mockito changed. Mockito v1.x has dependencies to hamcrest and objenesis whereas v2.x relies one byte-buddy and objenesis, so an other dependency don't make it worse. 

The problem you describe in your article is IMO not the added dependency that breaks the build, that can also happen if a library updates to a new version of a dependency . I think the root cause is the fact that you client relies on library (read mockito) version that has no final/GA release. If they use non final version they should be aware of changes. For the same reason many teams have a policy that only allows to use final releases.

**Comment by bric3** [comment]
I don't agree with these facts :

> Adding new dependency is a breaking change
> * New dependency is not present in corporate repo of one of your clients. His build breaks.

This argument completely disreagard dependency management tool like gradle or maven. Mockito is only distributed on bintray and maven central that are exactly made for getting those transitive dependencies.

> * New dependency is not allowed in corporate repo of your client (for example, due to poor reputation or security concerns). The client cannot use the new version.

With such policies, the importing library update itself is subject to these restrictions, and therefore updates are not allowed anyway until reviews approves changes of the importing library. This will happen regardless of the dependency being an external jar or being shaded.

e.g. license concerns won't be elimitated if the classes are just shaded.

> * New dependency is already declared by the client, but with different, incompatible version. Conflict resolution picks up later version causing failures at runtime.
> 
> Updating dependency could be a breaking change

This is indeed true and this is the only valid argument to me. Anyway the scope of this argument is limited because upgrading the importing library too can be in itself a breaking change (according to semver), e.g. mockito 2.x is incompatible with mockito 1.x.
This is indeed a problem with minor versions.

Updating a dependency is indeed something that could be breaking, but so could be the importing library.

**Comment by szczepiq** [comment]
@bric3, you're right, my first example in the use case is invalid. Thanks for pointing this out, I'll update the article.

@bric3, @ChristianSchwarz - you're right about the fact that major versions can carry breaking changes. However, including incompatible changes in subsequent release candidate is a poor engineering practice.

Release candidate by definition, is a software that is "ready to release" unless significant bugs are discovered. This means no new features, big changes, and most importantly, no incompatible changes. Consumers of release candidates need to be able to rely on this rule.

Thoughtful exceptions can always be made by the team, as we are doing :) I'll keep pushing for decent versioning practices. Thanks guys for feedback!!!

**Comment by bric3** [comment]
@szczepiq not there's still two items in #123, I think they can be safely removed, as we won't remove the serialVersionUid, and unincubated API have either _hatched out_ or stil incubating.

Also on mockito javadoc there's link that point to old website (like https://monkeyisland.pl/2009/01/13/subclass-and-override-vs-partial-mocking-vs-refactoring), I think we should only point to mockito wiki or github repo, not external website that can disappear anytime.

**Comment by szczepiq** [comment]
Done. 

>I think we should only point to mockito wiki or github repo

+1

**Comment by TimvdLippe** [comment]
I think we can publish on monday. The release seems very stable and we have yet another successful upgrade, this time at RxJava: https://github.com/ReactiveX/RxJava/pull/4636

**Comment by szczepiq** [comment]
+1. THANKS for great work :DDDD Pleasure doing projects together!!!

Great idea to migrate RxJava!

**Comment by TimvdLippe** [comment]
Published, w00h00. Upgrade today :tada:
